File: src/cli/document/list.ts
import type { Command } from "commander";
import type { LinearDocument } from "@linear/sdk";
import { getClient } from "../../lib/client.ts";
import { printError, printPaginated, resolvePageSize } from "../../lib/output.ts";
import { mapDocSummary } from "./map-doc-summary.ts";

export function registerList(document: Command): void {
  document
    .command("list")
    .description("List documents")
    .option("--project <project>", "Filter by project ID, slug, or name")
    .option("--creator <user>", "Filter by creator ID, name, or email")
    .option("--include-archived", "Include archived documents")
    .option("--limit <n>", "Limit results")
    .option("--cursor <token>", "Pagination cursor for next page")
    .action(async (opts: Record<string, string | undefined>) => {
      try {
        const client = getClient();
        const filter: Record<string, unknown> = {};

        if (opts.project) {
          filter.project = {
            or: [
              { id: { eq: opts.project } },
              { slugId: { eq: opts.project } },
              { name: { eqIgnoreCase: opts.project } },
            ],
          };
        }

        if (opts.creator) {
          filter.creator = {
            or: [
              { id: { eq: opts.creator } },
              { name: { eqIgnoreCase: opts.creator } },
              { displayName: { eqIgnoreCase: opts.creator } },
              { email: { eqIgnoreCase: opts.creator } },
            ],
          };
        }

        const results = await client.documents({
          first: resolvePageSize(opts),
          after: opts.cursor,
          filter:
            Object.keys(filter).length > 0 ? (filter as LinearDocument.DocumentFilter) : undefined,
          includeArchived: opts.includeArchived !== undefined ? true : undefined,
        });
        const items = await Promise.all(
          results.nodes.map((d) =>
            mapDocSummary(d as unknown as Parameters<typeof mapDocSummary>[0]),
          ),
        );
        printPaginated(items, results.pageInfo);
      } catch (err) {
        printError(err instanceof Error ? err.message : "List failed");
      }
    });
}

File: internal/cli/document/list.go
package document

import (
	"context"

	"github.com/spf13/cobra"

	"github.com/shhac/lin/internal/filters"
	"github.com/shhac/lin/internal/linear"
	"github.com/shhac/lin/internal/mappers"
	"github.com/shhac/lin/internal/output"
	"github.com/shhac/lin/internal/ptr"
)

func registerList(parent *cobra.Command) {
	var (
		project         string
		creator         string
		includeArchived bool
	)

	cmd := &cobra.Command{
		Use:   "list",
		Short: "List documents",
		Args:  cobra.NoArgs,
	}
	page := output.AddPageFlags(cmd)

	cmd.Run = func(cmd *cobra.Command, args []string) {
		client := linear.GetClient()
		ctx := context.Background()

		var filter *linear.DocumentFilter
		if project != "" || creator != "" {
			filter = &linear.DocumentFilter{}
			if project != "" {
				filter.Project = filters.BuildProjectFilter(project)
			}
			if creator != "" {
				filter.Creator = &linear.UserFilter{
					Or: []linear.UserFilter{
						{Id: &linear.IDComparator{Eq: ptr.To(creator)}},
						{Name: &linear.StringComparator{EqIgnoreCase: ptr.To(creator)}},
						{DisplayName: &linear.StringComparator{EqIgnoreCase: ptr.To(creator)}},
						{Email: &linear.StringComparator{EqIgnoreCase: ptr.To(creator)}},
					},
				}
			}
		}

		resp, err := linear.DocumentList(ctx, client, filter, page.Size(), page.Cursor(), ptr.TrueOrNil(includeArchived))
		if err != nil {
			output.HandleGraphQLError(err)
		}

		items := make([]any, len(resp.Documents.Nodes))
		for i, n := range resp.Documents.Nodes {
			items[i] = mappers.MapDocSummary(mappers.FromDocSummaryFields(n.DocSummaryFields))
		}

		output.PrintPage(items, resp.Documents.PageInfo.HasNextPage, resp.Documents.PageInfo.EndCursor)
	}

	cmd.Flags().StringVar(&project, "project", "", "Filter by project ID, slug, or name")
	cmd.Flags().StringVar(&creator, "creator", "", "Filter by creator ID, name, or email")
	cmd.Flags().BoolVar(&includeArchived, "include-archived", false, "Include archived documents")
	parent.AddCommand(cmd)
}

